refactor GarbageCollectionLogger so one instance per context

EdColeman opened the pull request:
Move GarbageCollectionLogger to ServerContext and start scheduled task in AbstractServer so that the is only one instance running for a server.

Bumping the the spotbugs version in #2824 highlighted a synchronization issue.  This change minimizes possibly having multiple instance of GarbageCollectionLogger running as well as making sure one runs per server instance.

dlmarion commented on server/base/src/main/java/org/apache/accumulo/server/AbstractServer.java:
```diff
@@ -54,6 +54,8 @@ protected AbstractServer(String appName, ServerOpts opts, String[] args) {
       // Server-side "client" check to make sure we're logged in as a user we expect to be
       context.enforceKerberosLogin();
     }
+    // start jvn gc logging
```
```suggestion
    // start jvm gc logging
```

EdColeman replied:
fixed in 80f5e35834

dlmarion commented on server/base/src/main/java/org/apache/accumulo/server/GarbageCollectionLogger.java:
```diff
@@ -22,109 +22,113 @@
-  }
 
+    if (maxIncreaseInCollectionTime > keepAliveTimeout) {
+      Halt.halt("Garbage collection may be interfering with lock keep-alive.  Halting.", -1);
```
I'm wondering if we need to still preemptively Halt the server since we have better exception handling in background threads now. If an OOME due to GC Overhead is thrown in a background thread it will halt the VM.

milleruntime replied:
I also wonder if we still need this halt for a number of reasons.

EdColeman replied:
I'd be happy to remove it - maybe as a separate issue?  Do you have test that demonstrates that it is not needed?

dlmarion replied:
No, just thinking out loud.

dlmarion commented on server/base/src/main/java/org/apache/accumulo/server/GarbageCollectionLogger.java:
```diff
@@ -22,109 +22,113 @@
+  }
 
-      boolean sawChange = false;
+  public synchronized void logGCInfo() {
```
I'm wondering if this method needs to be synchronized. Do we know what the lock in the previous version of the code is protecting? Is it the JMX MBeans? It looks like the only time this is called is either from the background thread, when the server is stopping, or when Halt is being invoked.

milleruntime replied:
There is a separate thread that gets started when the tserver is starting up but other threads call this method as well. I don't know why we have both the separate thread and external calls to this logger. The check for the static time was added because of an old complex issue: https://issues.apache.org/jira/browse/ACCUMULO-3084

dlmarion replied:
> I don't know why we have both the separate thread and external calls to this logger.

I'm understand why we have the background thread, but not why we are printing the JVM GC information when the server is stopped or halted.

> The check for the static time was added because of an old complex issue: https://issues.apache.org/jira/browse/ACCUMULO-3084

I see, the lock was used so that the `lastMemoryCheckTime` variable was not changed by another Thread. So, this code is trying to detect when this method is delayed due to GC. Wouldn't this cause an issue though in the case where:

1. The background thread is running, but delayed due to GC, and 
2. A call to `Halt` was issued, say by the Manager

milleruntime replied:
> this code is trying to detect when this method is delayed due to GC. Wouldn't this cause an issue

This is what I was thinking... A - It seems like this lock could cause problems. B - At best this thread _might_ have provided insight to Java GC on old versions of the JDK. C - Killing the tserver because of a guess at Java memory usage timing seems like a bad idea.

EdColeman replied:
In general - I think that this class can be removed in future versions in favor of metrics and recommending setting the appropriate  JVM options (-XX:HeapDumpOnOutOfMemoryError and/or -XX:OnOutOfMemeoryError) or whatever is appropriate for the given JVM.

However, the things that might prevent it from being removed or modified now:

- The halt in the class before an OOM is thrown may allow for a more orderly shutdown.  We should be resistant to sudden failures, but a few moments to flush / close thing might still speed recovery - but not sure of this.
- Users may be expecting these messages and/or not have the necessary metrics captured. 
- I am not sure right now if the JVM metrics alone would be sufficient, or if some variation of the values gathered / logged here should be added.

Removal / metrics changes could be another follow-on for this class.  The goal with this PR was to be an incremental improvement to resolve the synchronization issue flagged by spotbugs.  While this current iteration may not be ideal, it should at least be correct if there are concurrent log attempts whiteout significantly changing the behavior.

milleruntime commented on server/base/src/main/java/org/apache/accumulo/server/ServerContext.java:
```diff
@@ -132,6 +134,7 @@ private ServerContext(ServerInfo info) {
     securityOperation =
         memoize(() -> new AuditedSecurityOperation(this, SecurityOperation.getAuthorizor(this),
             SecurityOperation.getAuthenticator(this), SecurityOperation.getPermHandler(this)));
+    jvmGcLogger = memoize(() -> new GarbageCollectionLogger(this));
```
The local name `jvmGcLogger` is better but I think the name of this class is confusing. Some reference to Java or the JVM in the name would greatly help to differentiate it from the Accumulo GC classes. Something like `JVMGcLogger` or `JavaMemoryLogger` or `JavaGcLogger` would be a lot better.

EdColeman replied:
Renamed in 80f5e35834 to `JvmGcLogger`

milleruntime commented:
While I appreciate the attempt to improve the use of this class, I really think the logic of this class should exist in the `TabletServerResourceManager`. I am not sure how exactly it would be incorporated with the other memory checks but we have a `memoryGuardThread` already running in the tserver. This thread is more geared towards `Tablet` memory usage but I feel like the Java GC checking thread here should be running with the other resource management of the tserver. The `GarbageCollectionLogger` class feels like it was just created quickly on the fly to try for debugging a complex production issue and just thrown into the tserver.

dlmarion commented:
> I really think the logic of this class should exist in the TabletServerResourceManager

The JVM GC Logger is used by all server processes, not just the TabletServer.

EdColeman commented:
I believe that I have addressed the issues as much as can be done without a total refactor.  I think it represents an incremental improvement of the current code and larger changes should be handled as a separate issue.  Or, just abandon this can stay with what we currently have?

milleruntime commented:
> I believe that I have addressed the issues as much as can be done without a total refactor. I think it represents an incremental improvement of the current code and larger changes should be handled as a separate issue. Or, just abandon this can stay with what we currently have?

I am not sure this change is worth the benefit, especially if there is a good chance this class is going to be removed in the future. I would be more in favor of a change that makes it easier to remove the class. Now that I think about it, these changes might make it more difficult to drop the class.

EdColeman commented:
How does this make future changes harder?  

  - It moves the instantiation of instances to one place. 
  - Instead of multiple instances running in scheduled executors there is one.
  - The instance is created in the AbstractServer code so that any service gets an instance without needed to add code that is not obvious that we use.

milleruntime commented:
> How does this make future changes harder?

I didn't say that. I said it makes it more difficult to drop the class. The class is being renamed, moved and its usage refactored. It is being introduced into new classes where it wasn't referenced previously. This PR increases the complexity of the class and its usage. I am advocating for doing nothing and just removing the class. For example... GarbageCollectionLogger is now only referenced 11 times, while the new class is 29.

EdColeman commented:
The rename was a suggestion by @milleruntime 

```
...I think the name of this class is confusing. Some reference to Java or the JVM in the name would greatly help to differentiate it from the Accumulo GC classes. Something like JVMGcLogger or JavaMemoryLogger or JavaGcLogger would be a lot better.
```

milleruntime commented:
> The rename was a suggestion by @milleruntime
> 
> ```
> ...I think the name of this class is confusing. Some reference to Java or the JVM in the name would greatly help to differentiate it from the Accumulo GC classes. Something like JVMGcLogger or JavaMemoryLogger or JavaGcLogger would be a lot better.
> ```

Yes, the name is confusing.

ctubbsii commented:
I'm also in favor of removing this class, and share @milleruntime 's concern about introducing it in new classes where it wasn't used before. However, I do find the new name less confusing, if we end up keeping it. I also like making it a singleton, if we're keeping it around. We don't need more than one of these per process.

I think we should push this off to 3.0, at least, and decide then what we want to do with it.

EdColeman commented:
This PR will be superseded by the changes with https://github.com/apache/accumulo/pull/3161

ctubbsii commented:
> This PR will be superseded by the changes with #3161

In that case, I'm closing this, so we can focus on reviewing that one.
